src/main/java/net/imglib2/algorithm/fill/FloodFill.java

```diff
@@ -0,0 +1,207 @@
+     *
+     * @param source input
+     * @param target {@link RandomAccessible} to be written into. May be the same as input.
+     * @param seed Start flood fill at this location.
+     * @param seedLabel Immutable. Reference value of input at seed location.
+     * @param fillLabel Immutable. Value to be written into valid flood fill locations.
+     * @param <T> T implements {@link Comparable<T>}.
+     * @param <U> U implements {@link Type<U>} and {@link Comparable<U>}.
+     */
+    public static < T extends Comparable< T >, U extends Type< U > & Comparable< U > > void fill(
+            final RandomAccessible< T > source,
+            final RandomAccessible< U > target,
+            final Localizable seed,
+            final T seedLabel,
+            final U fillLabel,
+            final Shape shape
+    )
+    {
+        fill( source, target, seed, seedLabel, fillLabel, shape, new ComparableComparator<T, U>(), new TypeWriter<U>() );
+    }
+
+
+    /**
+     *
+     * Iterative n-dimensional flood fill for arbitrary neighborhoods: Starting at seed location, write fillLabel
+     * into target at current location and continue for each pixel in neighborhood defined by shape if neighborhood
+     * pixel is in the same connected component and fillLabel has not been written into that location yet (comparator
+     * evaluates to 0).
+     *
+     * @param source input
+     * @param target {@link RandomAccessible} to be written into. May be the same as input.
+     * @param seed Start flood fill at this location.
+     * @param seedLabel Immutable. Reference value of input at seed location.
+     * @param fillLabel Immutable. Value to be written into valid flood fill locations.
+     * @param shape Defines neighborhood that is considered for connected components, e.g. {@link net.imglib2.algorithm.neighborhood.DiamondShape}
+     * @param comparator Returns 0 if pixel has not been visited yet and should be written into. Returns non-zero if target pixel has been visited
+     *                   or source pixel is not part of the same connected component. In the most trivial case, comparator will return 0
+     *                   if and only if source at the current location is equal to seedLabel and target at the current location is different
+     *                   from fillLabel, cf {@link ComparableComparator}.
+     * @param writer Defines how fillLabel is written into target at current location.
+     * @param <T> No restrictions on T. Appropriate comparator is the only requirement.
+     * @param <U> No restrictions on U. Appropriate comparator and writer is the only requirement.
+     */
+    public static < T, U > void fill(
+            final RandomAccessible< T > source,
+            final RandomAccessible< U > target,
+            final Localizable seed,
+            final T seedLabel,
+            final U fillLabel,
+            final Shape shape,
+            final Comparator< Pair< T, U > > comparator,
```

Comment:
The more I think about it, the more I think, we should not use the `Comparator` interface as it does not express what we are actually interested in. We are interested only in a true/false decision. Returning an order does not provide any useful information and makes things less intuitive. Ideally, we would have the same interface as `Comparator` that returns `Boolean` instead of `Integer`. Java 8's [BiFunction](https://docs.oracle.com/javase/8/docs/api/java/util/function/BiFunction.html) seems like a reasonable choice. Instead of `Comparator< Pair< T, U > > comparator` we would then pass `BiFunction< Pair< T, U >, Pair< T, U >, Boolean > checkLocation` (or better name if we find one). Using this interface would conform more with what we actually use the current `comparator` for.